core/src/main/java/org/apache/calcite/sql/fun/SqlJsonInsertFunction.java

```diff
@@ -0,0 +1,108 @@
+ * limitations under the License.
+ */
+package org.apache.calcite.sql.fun;
+
+import org.apache.calcite.rel.type.RelDataType;
+import org.apache.calcite.sql.SqlCall;
+import org.apache.calcite.sql.SqlCallBinding;
+import org.apache.calcite.sql.SqlFunction;
+import org.apache.calcite.sql.SqlFunctionCategory;
+import org.apache.calcite.sql.SqlKind;
+import org.apache.calcite.sql.SqlOperandCountRange;
+import org.apache.calcite.sql.SqlWriter;
+import org.apache.calcite.sql.type.ReturnTypes;
+import org.apache.calcite.sql.type.SqlOperandCountRanges;
+import org.apache.calcite.sql.type.SqlOperandTypeChecker;
+import org.apache.calcite.sql.type.SqlTypeUtil;
+import org.apache.calcite.sql.validate.SqlValidator;
+
+import java.util.Locale;
+
+import static org.apache.calcite.util.Static.RESOURCE;
+
+/**
+ * The <code>JSON_INSERT</code> function.
+ */
+public class SqlJsonInsertFunction extends SqlFunction {
+  public SqlJsonInsertFunction() {
+    super("JSON_INSERT", SqlKind.OTHER_FUNCTION, ReturnTypes.VARCHAR_2000, null,
+        null, SqlFunctionCategory.SYSTEM);
+  }
+
+  @Override public SqlOperandCountRange getOperandCountRange() {
+    return SqlOperandCountRanges.from(3);
+  }
+
+  @Override protected void checkOperandCount(SqlValidator validator,
+      SqlOperandTypeChecker argType, SqlCall call) {
+    assert (call.operandCount() >= 3) && (call.operandCount() % 2 == 1);
+  }
+
+  @Override public boolean checkOperandTypes(SqlCallBinding callBinding,
+      boolean throwOnFailure) {
+    final int count = callBinding.getOperandCount();
+    for (int i = 1; i < count; i += 2) {
+      RelDataType nameType = callBinding.getOperandType(i);
+      if (!SqlTypeUtil.inCharFamily(nameType)) {
+        if (throwOnFailure) {
+          throw callBinding.newError(RESOURCE.expectedCharacter());
+        }
+        return false;
+      }
+      if (nameType.isNullable()) {
+        if (throwOnFailure) {
+          throw callBinding.newError(
+              RESOURCE.argumentMustNotBeNull(
+                  callBinding.operand(i).toString()));
+        }
+        return false;
+      }
+    }
+    return true;
+  }
+
+  @Override public String getSignatureTemplate(int operandsCount) {
+    assert operandsCount % 2 == 1;
+    StringBuilder sb = new StringBuilder();
+    sb.append("{0}(");
+    for (int i = 1; i < operandsCount; i++) {
+      sb.append(String.format(Locale.ROOT, "{%d} ", i + 1));
+    }
+    sb.append("{1})");
+    return sb.toString();
+  }
+
+  @Override public void unparse(SqlWriter writer, SqlCall call, int leftPrec,
+      int rightPrec) {
+    assert call.operandCount() % 2 == 1;
+    final SqlWriter.Frame frame = writer.startFunCall(getName());
+    call.operand(0).unparse(writer, 0, 0);
+    SqlWriter.Frame listFrame = writer.startList("", "");
+    writer.sep(",");
+    for (int i = 1; i < call.operandCount(); i += 2) {
+      writer.sep(",");
+      writer.keyword("KEY");
```

Comment:
I remember that the keywords `KEY`/`VALUE` are added for JSON_OBJECT(AGG) functions. I don't think MySQL's JSON functions have similar syntax. Could you please double check (the same to JSON_SET, JSON_REPLACE)? Also, if you copied codes from another place, please make sure every line you copied is reasonable for staying at the new place.